### PR TITLE
feat(cmds): allow to set the configuration file path

### PR DESCRIPTION
**Pull request by schomatis**
First iteration, still incomplete. Review needed at this point to move forward.

Depends on https://github.com/ipfs/go-ipfs-config/pull/162.

Closes https://github.com/ipfs/go-ipfs/issues/8067.

### CODE REVIEW

File: core/commands/root.go [hunk 1]
```diff
@@ -19,6 +19,17 @@ var log = logging.Logger("core/commands")
...
+	// FIXME(BLOCKING): Revisit these names. They were chosen for the review
+	//  as they make explicit whether they signal a directory or a file (and
+	//  we needed to differentiate them from the old "config" option).
+	RepoDirOption    = "repo-dir"
+	ConfigFileOption = "config-file"
+	// FIXME(BLOCKING): Decide what to do when the user sets the deprecated
+	//  (and incorrectly named) "config" option, either:
+	//  * fail entirely
+	//  * issue a warning but process it as if it were the old equivalent RepoDirOption
+	//    (and fail if either of the new options that replace it are set)
```
Comment (guseggert) [inline core/commands/root.go:31]: Do we have some precedence for how we've done this in the past? Abstractly I'd prefer this option but we should probably do the least surprising thing, whatever that is (if it's to just make the breaking change, then IMO we should just do it).

File: repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go [hunk 2]
```diff
@@ -287,7 +287,9 @@ func readIpfsConfig(repoRoot *string) (bootstrap []string, peers []peer.AddrInfo
 		return
 	}
 
-	cfgPath, err := config.Filename(*repoRoot)
+	// FIXME(BLOCKING): Decide if we need to add the config file path as
+	//  an extra variable in IpfsFetcher (and extend its constructor).
+	cfgPath, err := config.Filename(*repoRoot, "")
```
Comment (guseggert) [inline repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go:292]: Yeah we need this for migrations to work correctly right? AFAIU, nobody consumes this directly except us in go-ipfs and ipfs-update so we should be ok to break it.

File: repo/fsrepo/migrations/migrations.go [hunk 3]
```diff
@@ -120,7 +120,8 @@ func ReadMigrationConfig(repoRoot string) (*config.Migration, error) {
 		Migration config.Migration
 	}
 
-	cfgPath, err := config.Filename(repoRoot)
+	// FIXME(BLOCKING): Decide if we need to add the config file path as an argument (API breaking).
+	cfgPath, err := config.Filename(repoRoot, "")
```
Comment (guseggert) [inline repo/fsrepo/migrations/migrations.go:124]: yeah same rationale as IpfsFetcher...fine to break this

File: core/commands/root.go (hunk 1 above)
Comment (schomatis) [reply core/commands/root.go:31]: I don't know; will need someone else to make that call.

**Comment by BigLep** [comment]
@guseggert : can you do the final review/merge so we can close this out?

File: core/commands/root.go [hunk 4]
```diff
@@ -19,6 +19,17 @@ var log = logging.Logger("core/commands")
 var ErrNotOnline = errors.New("this command must be run in online mode. Try running 'ipfs daemon' first")
 
 const (
+	// FIXME(BLOCKING): Revisit these names. They were chosen for the review
+	//  as they make explicit whether they signal a directory or a file (and
+	//  we needed to differentiate them from the old "config" option).
```
Comment (guseggert) [inline core/commands/root.go:24]: remove now that we have clarity?

File: core/commands/root.go (hunk 1 above)
Comment (guseggert) [reply core/commands/root.go:31]: now that we are running with "deprecate", can we remove this comment?

File: go.mod [hunk 5]
```diff
@@ -31,7 +31,8 @@ require (
 	github.com/ipfs/go-ipfs-blockstore v1.1.2
 	github.com/ipfs/go-ipfs-chunker v0.0.5
 	github.com/ipfs/go-ipfs-cmds v0.6.0
-	github.com/ipfs/go-ipfs-config v0.19.0
+	// FIXME(BLOCKING): Update to release tag once https://github.com/ipfs/go-ipfs-config/pull/162 lands.
+	github.com/ipfs/go-ipfs-config v0.19.1-0.20220221223623-ea7b3a0b7fba
```
Comment (guseggert) [inline go.mod:35]: update

File: repo/fsrepo/fsrepo.go [hunk 6]
```diff
@@ -185,13 +197,19 @@ func open(repoPath string) (repo.Repo, error) {
...
+func newFSRepo(rpath string, userConfigFilePath string) (*FSRepo, error) {
 	expPath, err := homedir.Expand(filepath.Clean(rpath))
 	if err != nil {
 		return nil, err
 	}
 
-	return &FSRepo{path: expPath}, nil
+	configFilePath, err := config.Filename(rpath, userConfigFilePath)
+	if err != nil {
+		// FIXME: Personalize this when the user config path is "".
```
Comment (guseggert) [inline repo/fsrepo/fsrepo.go:208]: Do we need to do this?

File: repo/fsrepo/fsrepo.go [hunk 7]
```diff
@@ -214,7 +232,8 @@ func ConfigAt(repoPath string) (*config.Config, error) {
 	packageLock.Lock()
 	defer packageLock.Unlock()
 
-	configFilename, err := config.Filename(repoPath)
+	// FIXME(BLOCKING): Decide if we need to add the config file path as an argument (API breaking).
```
Comment (guseggert) [inline repo/fsrepo/fsrepo.go:235]: remove?

File: repo/fsrepo/fsrepo.go [hunk 8]
```diff
@@ -238,7 +258,8 @@ func initConfig(path string, conf *config.Config) error {
 	if configIsInitialized(path) {
 		return nil
 	}
-	configFilename, err := config.Filename(path)
+	// FIXME: Depends on what's decided on Init.
```
Comment (guseggert) [inline repo/fsrepo/fsrepo.go:261]: remove?

File: repo/fsrepo/fsrepo.go [hunk 9]
```diff
@@ -224,7 +243,8 @@ func ConfigAt(repoPath string) (*config.Config, error) {
 // configIsInitialized returns true if the repo is initialized at
 // provided |path|.
 func configIsInitialized(path string) bool {
-	configFilename, err := config.Filename(path)
+	// FIXME: Depends on what's decided on Init.
```
Comment (guseggert) [inline repo/fsrepo/fsrepo.go:246]: remove?

File: repo/fsrepo/fsrepo.go [hunk 10]
```diff
@@ -273,6 +294,8 @@ func initSpec(path string, conf map[string]interface{}) error {
 
 // Init initializes a new FSRepo at the given path with the provided config.
 // TODO add support for custom datastores.
+// FIXME(BLOCKING): Decide if we want to allow the user to set a different config
```
Comment (guseggert) [inline repo/fsrepo/fsrepo.go:297]: remove?

File: repo/fsrepo/fsrepo.go (hunk 6 above)
Comment (schomatis) [reply repo/fsrepo/fsrepo.go:208]: Not blocking, just a normal FIXME, can remove if you want.

**Comment by lidel** [comment]
- bad news:  https://github.com/ipfs/go-ipfs-config/pull/162 was not  in https://github.com/ipfs/go-ipfs/pull/8756
- good news: config lives in `./config` dir of go-ipfs repo, so you no longer need to open two PRs

@schomatis  I think you need to rebase this PR and then apply changes from https://github.com/ipfs/go-ipfs-config/pull/162  into ./config

**Review by guseggert — APPROVED** [review]

**Comment by guseggert** [comment]
Can you add a sharness test covering these new flags?

**Comment by BigLep** [comment]
@guseggert : it says you approved the changes but I see you've also requested a sharness test.  To make it clear, do you want to adjust your review to "request changes"?

**Review by guseggert — CHANGES_REQUESTED** [review]
Can you add a sharness test covering these new flags?

**Comment by schomatis** [comment]
Not sure what are the `ci/circleci: go-ipfs-http-client` fails about.

**Review by guseggert — APPROVED** [review]

**Comment by BigLep** [comment]
@schomatis : can you rebase please?  That will fix CI and then we can merge.

**Pull request merged**